src/hotspot/share/memory/metaspace/freeChunkList.hpp

```diff
@@ -0,0 +1,272 @@
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ *
+ */
+
+#ifndef SHARE_MEMORY_METASPACE_FREECHUNKLIST_HPP
+#define SHARE_MEMORY_METASPACE_FREECHUNKLIST_HPP
+
+#include "memory/allocation.hpp"
+#include "memory/metaspace/chunklevel.hpp"
+#include "memory/metaspace/counters.hpp"
+#include "memory/metaspace/metachunk.hpp"
+#include "memory/metaspace/metachunkList.hpp"
+
+class outputStream;
+
+namespace metaspace {
+
+// This is the free list underlying the ChunkManager.
+//
+// Chunks are kept in a vector of double-linked double-headed lists
+//  (using Metachunk::prev/next). One list per chunk level exists.
+//
+// Chunks in these lists are roughly ordered: uncommitted chunks
+//  are added to the back of the list, fully or partially committed
+//  chunks to the front.
+//
+// (Small caveat: commit state of a chunk may change as a result of
+//  actions on neighboring chunks, if the chunk is smaller than a commit
+//  granule and therefore shares its granule with neighbors. So it may change
+//  after the chunk has been added to the list.
+//  It will never involuntarily uncommit: only chunks >= granule size are uncommitted.
+//  But it may get involuntarily committed if an in-granule neighbor is committed and
+//  causes committing of the whole granule.
+//  In practice this is not a big deal; it has very little consequence.)
+//
+// Beyond adding at either front or at back, we do not sort on insert, since the
+//  insert path is used during Metaspace reclamation which may happen at GC pause.
+//
+// During retrieval (at class loading), we search the list for a chunk of at least
+//  n committed words to satisfy the caller requested committed word size. We stop
+//  searching at the first fully uncommitted chunk.
+//
+// Note that even though this is an O(n) search, in practice this is not a problem:
+//  - in all likelihood the requested commit word size is way smaller than even a single
+//    commit granule, so 99% of all searches would end at the first chunk (which is either
+//    uncommitted or committed to at least one commit granule size).
+//  - in all likelihood chunks, when added to this list, are either fully committed
+//    or fully uncommitted (see Settings::uncommit_on_return_min_word_size()).
+//
+// Should we ever encounter situations where the O(n) search is a bottleneck, this
+//  structure can easily be optimized (e.g. a BST). But for now lets keep this simple.
+
+class FreeChunkList {
```

Comment:
Yes, this could possibly be done. Its a bit of a matter of taste. In this case note that FreeChunkList orders chunks by commit state on insert, and does other stuff particular to that list, so you will need to revive that functionality. As it is now I find this coding clear and easy to read. Any solution we arrive at after factoring out should not be worse.

Reply:
I have not yet looked closely at freeChunkList.[ch]pp. Maybe `FreeChunkList` isn't a good candidate for doing it. I still think there is potential for reuse. I'm ok with postponing it though.